network_wrangler/roadway/network.py

```diff
@@ -611,6 +646,242 @@ def delete_nodes(
+            "X": split_point.x,
+            "Y": split_point.y,
+        }
+
+        # Copy additional attributes from node A if they exist
+        node_a_data = self.nodes_df.loc[self.nodes_df["model_node_id"] == orig_link["A"]].iloc[0]
+        # Copy all columns except geometry, X, Y, and model_node_id
+        for col in self.nodes_df.columns:
+            if (
+                col not in ["model_node_id", "X", "Y", "geometry", "osm_node_id"]
+                and col in node_a_data.index
+                and pd.notna(node_a_data[col])
+            ):
+                new_node_data[col] = node_a_data[col]
+
+        # Create DataFrame for new node and add it
+        new_node_df = pd.DataFrame([new_node_data])
+        self.add_nodes(new_node_df)
+
+        # Split the geometry
+        # Create a small circle around the split point to ensure clean split
+        split_geom = split_point.buffer(0.00001)  # Small buffer in degrees
+        split_result = split(geometry, split_geom)
+
+        if len(split_result.geoms) >= MIN_SPLIT_SEGMENTS:
+            geom1 = split_result.geoms[0]
+            geom2 = split_result.geoms[-1]
+        else:
+            # Fallback: manually create two linestrings
+            coords = list(geometry.coords)
+            split_idx = int(len(coords) * fraction)
+            if split_idx == 0:
+                split_idx = 1
+            elif split_idx >= len(coords):
+                split_idx = len(coords) - 1
+
+            # Insert the split point at the right position
+            coords_1 = [*coords[:split_idx], (split_point.x, split_point.y)]
+            coords_2 = [(split_point.x, split_point.y), *coords[split_idx:]]
+
+            geom1 = LineString(coords_1)
+            geom2 = LineString(coords_2)
+
+        # Create two new links
+        # Find the next available model_link_ids
+        max_link_id = self.links_df["model_link_id"].max()
+        new_link1_id = max_link_id + 1
+        new_link2_id = max_link_id + 2
+
+        # First link: A to new node
+        link1_data = orig_link.to_dict()
+        link1_data["model_link_id"] = new_link1_id
+        link1_data["B"] = new_model_node_id
+        link1_data["geometry"] = geom1
+        link1_data["distance"] = (
+            orig_link.get("distance", 0) * fraction if "distance" in orig_link else None
+        )
+        # Clear shape_id as we're creating new geometry
+        link1_data["shape_id"] = None
+
+        # Second link: new node to B
+        link2_data = orig_link.to_dict()
+        link2_data["model_link_id"] = new_link2_id
+        link2_data["A"] = new_model_node_id
+        link2_data["geometry"] = geom2
+        link2_data["distance"] = (
+            orig_link.get("distance", 0) * (1 - fraction) if "distance" in orig_link else None
+        )
+        # Clear shape_id as we're creating new geometry
+        link2_data["shape_id"] = None
+
+        # Handle reverse link if requested
```

Comment:
It seems like it would make sense to encapsulate this whole thing in a funct and call it once on forward and once on reverse. As it is it is really verbose.